src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala

```diff
@@ -0,0 +1,724 @@
+    val colPartitions = partitioner.colPartitions
+    val truncatedBlockRow = lm.rows / blockSize
+    val truncatedBlockCol = lm.cols / blockSize
+    val excessRows = lm.rows % blockSize
+    val excessCols = lm.cols % blockSize
+    val indices = for {
+      i <- 0 until rowPartitions
+      j <- 0 until colPartitions
+    } yield (i, j)
+    val blocks = sc.parallelize(indices).map { case (i, j) =>
+      val iOffset = i * blockSize
+      val jOffset = j * blockSize
+      val rowsInBlock = if (i == truncatedBlockRow) excessRows else blockSize
+      val colsInBlock = if (j == truncatedBlockCol) excessCols else blockSize
+      val a = new Array[Double](rowsInBlock * colsInBlock)
+      var jj = 0
+      while (jj < colsInBlock) {
+        var ii = 0
+        while (ii < rowsInBlock) {
+          a(jj * rowsInBlock + ii) = lmBc.value(iOffset + ii, jOffset + jj)
+          ii += 1
+        }
+        jj += 1
+      }
+      ((i, j), new BDM(rowsInBlock, colsInBlock, a))
+    }.partitionBy(partitioner)
+    new BlockMatrix(blocks, blockSize, lm.rows, lm.cols)
+  }
+
+  def from(irm: IndexedRowMatrix): M =
+    from(irm, defaultBlockSize)
+
+  def from(irm: IndexedRowMatrix, blockSize: Int): M =
+    irm.toHailBlockMatrix(blockSize)
+
+  def map4(f: (Double, Double, Double, Double) => Double)(a: M, b: M, c: M, d: M): M =
+    a.map4(b, c, d, f)
+
+  def map2(f: (Double, Double) => Double)(l: M, r: M): M =
+    l.map2(r, f)
+
+  private class PairWriter(var i: Int, var j: Int) extends Writable {
+    def this() {
+      this(0, 0)
+    }
+
+    def write(out: DataOutput) {
+      out.writeInt(i)
+      out.writeInt(j)
+    }
+
+    def readFields(in: DataInput) {
+      i = in.readInt()
+      j = in.readInt()
+    }
+  }
+
+  private class MatrixWriter(var rows: Int, var cols: Int, var a: Array[Double]) extends Writable {
+    def this() {
+      this(0, 0, null)
+    }
+
+    def write(out: DataOutput) {
+      out.writeInt(rows)
+      out.writeInt(cols)
+      var i = 0
+      while (i < rows * cols) {
+        out.writeDouble(a(i))
+        i += 1
+      }
+    }
+
+    def readFields(in: DataInput) {
+      rows = in.readInt()
+      cols = in.readInt()
+      a = new Array[Double](rows * cols)
+      var i = 0
+      while (i < rows * cols) {
+        a(i) = in.readDouble()
+        i += 1
+      }
+    }
+
+    def toDenseMatrix(): BDM[Double] = {
+      new BDM[Double](rows, cols, a)
+    }
+  }
+
+  private val metadataRelativePath = "/metadata.json"
+  private val matrixRelativePath = "/matrix"
+
+  /**
+    * Writes the matrix {@code m} to a Hadoop sequence file at location {@code
+    * uri}.
+    *
+    **/
+  def write(dm: M, uri: String) {
+    val hadoop = dm.blocks.sparkContext.hadoopConfiguration
+    hadoop.mkDir(uri)
+
+    dm.blocks.map { case ((i, j), lm) =>
+      (new PairWriter(i, j), new MatrixWriter(lm.rows, lm.cols, lm.toArray))
+    }
+      .saveAsSequenceFile(uri + matrixRelativePath)
```

Comment:
Can I ask, what happened to the code to preserve partitioning?  I don't see another PR for it.

Reply:
It's in #2276, but I closed it until I've abstracted it based on your suggestion (next week)